pkg/registry/core/service/storage/storage.go

```diff
@@ -240,3 +291,163 @@ func (r *GenericREST) defaultOnReadService(service *api.Service) {
 		}
 	}
 }
+
+func (r *GenericREST) afterDelete(obj runtime.Object, options *metav1.DeleteOptions) {
+	svc := obj.(*api.Service)
+
+	// Normally this defaulting is done automatically, but the hook (Decorator)
+	// is called at the end of this process, and we want the fully-formed
+	// object.
+	r.defaultOnReadService(svc)
+
+	// Only perform the cleanup if this is a non-dryrun deletion
+	if !dryrun.IsDryRun(options.DryRun) {
+		// It would be better if we had the caller context, but that changes
+		// this hook signature.
+		ctx := genericapirequest.WithNamespace(genericapirequest.NewContext(), svc.Namespace)
+		// TODO: This is clumsy.  It was added for fear that the endpoints
+		// controller might lag, and we could end up rusing the service name
+		// with old endpoints.  We should solve that better and remove this, or
+		// else we should do this for EndpointSlice, too.
```

Comment:
I don't think slices has this problem, since it reconciles on slices events, when the controller restarts it will requeue the orphan endpoint and it should be deleted ... I didn't verify it though